### ISSUE TITLE
Copy & Paste

### ISSUE THREAD

**Issue by RyanRiffle** [RyanRiffle/Poe#3]
Provide copy and paste support. At a minimum be able to copy and paste text only without formatting.

**Comment by MrSchism** [comment]
Might be worth looking into the HTML5 Clipboard API.  It doesn't work as well as it could, but it works in most instances on current browsers.  It's better than nothing (until we get better support or resort to ZeroClipboard).

Reference: http://www.w3.org/TR/clipboard-apis/

**Comment by RyanRiffle** [comment]
@MrSchism In order for copy paste to be implemented, first text selection really needs to be handled by Poe itself, because when tables and images get added, stock js ranges will not be too great to work with.

**Comment by andersevenrud** [comment]
I have looked into this in the past regarding OS.js. Since clipboard is considered unsafe, most browsers does not allow this behavior... except IE (shocker... right?). There are Flash solutions for this, but even that is slowly becoming unsusable because of new security policies regarding Flash.

https://github.com/mojombo/clippy

**Comment by andersevenrud** [comment]
But you can however create custom keyboard bindings for this pretty easily. Global copy/paste via OS clipboard however is not that pretty.

**Comment by RyanRiffle** [comment]
@andersevenrud Yea copy and pasting is horrible when it comes to web browsers. Its been an issue for years, I wonder why the vendors have not done something about it?

**Comment by andersevenrud** [comment]
@RyanRiffle It's quite simple. It's considered harmful. If you had complete access to the clipboard you could insert malicious links and/or code (if the implementation has a buffer overflow or something, things go wrong fast!).

**Comment by MrSchism** [comment]
Yeah.  Luckily they're working on improving HTML5 APIs to replace the Flash solutions... but there's still no real solution.

**Comment by RyanRiffle** [comment]
Copy and paste will not be hard to implement now. Text selection now works due to the recent rewrite of the project and change from Coffeescript to Javascript. So now it is just a matter of getting the elements that are selected and detaching them from the DOM and inserting them when necessary. Getting all of the selected words is already in the API as well so piece of cake!

**Comment by RyanRiffle** [comment]
As a note: The desktop version of Poe will possibly be able to support Copy + Paste using the system clipboard, however the web version will need something more as stated above. At the time of this writing Copy has been implemented but the inverse Paste has not been worked out completely. This is only for internal copy + paste at the moment.
